Aya doesn't accept the application if there is a solve-later meta

```aya
def how {A B : Type} (f : A -> B) (a : A) : B => f a
def what {A B : Type} (a : A) (f : A -> B) : B => how (fn f => f a) f
```

<details>
<summaryAya says:</summary>

```plain
Exception in thread "main" java.lang.AssertionError: Failed with `class org.aya.tyck.error.BadTypeError`: In file <baka>:2:65 ->

  1 |   def how0 {A B : Type} (f : A -> B) (a : A) : B => f a
  2 |   def what0 {A B : Type} (a : A) (f : A -> B) : B => how0 (fn f => f a) f
    |                                                                    ^-^

Error: Unable to apply the expression
         f a
       because the type of what you applied is not a Pi/Path type, but instead:
         A -> B

at (119-121) [2,65-2,67]
```

</details>

I found that the type of `f` is a `MetaCall` while checking the application, so Aya rejects it, but it will be solved after the unification of `f` and `(a : A)`.

--- a/base/src/main/java/org/aya/normalize/Finalizer.java
+++ b/base/src/main/java/org/aya/normalize/Finalizer.java
@@ -2,6 +2,7 @@
 // Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
 package org.aya.normalize;
 
+import kala.collection.mutable.MutableList;
 import kala.collection.mutable.MutableSinglyLinkedList;
 import org.aya.normalize.error.UnsolvedLit;
 import org.aya.normalize.error.UnsolvedMeta;
@@ -9,6 +10,7 @@ import org.aya.syntax.core.term.MetaPatTerm;
 import org.aya.syntax.core.term.Term;
 import org.aya.syntax.core.term.call.MetaCall;
 import org.aya.syntax.core.term.repr.MetaLitTerm;
+import org.aya.syntax.ref.MetaVar;
 import org.aya.tyck.TyckState;
 import org.aya.tyck.tycker.Problematic;
 import org.aya.tyck.tycker.Stateful;
@@ -33,10 +35,11 @@ public interface Finalizer {
   }
 
   record Zonk<T extends Problematic & Stateful>(
-    @NotNull T delegate, @NotNull MutableSinglyLinkedList<Term> stack
+    @NotNull T delegate, @NotNull MutableSinglyLinkedList<Term> stack,
+    @NotNull MutableList<MetaVar> alreadyReported
   ) implements Finalizer, Stateful, Problematic {
     public Zonk(@NotNull T delegate) {
-      this(delegate, MutableSinglyLinkedList.create());
+      this(delegate, MutableSinglyLinkedList.create(), MutableList.create());
     }
     @Override public @NotNull TyckState state() { return delegate.state(); }
     @Override public @NotNull Reporter reporter() { return delegate.reporter(); }
@@ -45,10 +48,13 @@ public interface Finalizer {
       var result = doZonk(term);
       // result shall not be MetaPatTerm
       switch (result) {
-        case MetaCall meta when !meta.ref().isUser() -> fail(new UnsolvedMeta(stack.view()
-          .drop(1)
-          .map(this::freezeHoles)
-          .toImmutableSeq(), meta.ref().pos(), meta.ref().name()));
+        case MetaCall(var ref, _) when !ref.isUser() && !alreadyReported.contains(ref) -> {
+          alreadyReported.append(ref);
+          fail(new UnsolvedMeta(stack.view()
+            .drop(1)
+            .map(this::freezeHoles)
+            .toImmutableSeq(), ref.pos(), ref.name()));
+        }
         case MetaLitTerm mlt -> fail(new UnsolvedLit(mlt));
         default -> {
         }

--- a/base/src/main/java/org/aya/tyck/ArgsComputer.java
+++ b/base/src/main/java/org/aya/tyck/ArgsComputer.java
@@ -7,6 +7,7 @@ import kala.collection.mutable.MutableStack;
 import org.aya.generic.term.DTKind;
 import org.aya.generic.term.SortKind;
 import org.aya.syntax.concrete.Expr;
+import org.aya.syntax.core.Closure;
 import org.aya.syntax.core.term.*;
 import org.aya.syntax.core.term.call.ClassCall;
 import org.aya.syntax.core.term.call.MetaCall;
@@ -91,6 +92,9 @@ public class ArgsComputer {
           var argJdg = tycker.synthesize(arg.arg());
           var codMeta = new MetaVar(ref.name() + "_cod", ref.pos(), ref.ctxSize() + 1, req, false);
           var cod = new MetaCall(codMeta, metaArgs.appended(argJdg.wellTyped()));
+          tycker.solve(ref, new DepTypeTerm(DTKind.Pi,
+            argJdg.type(), new Closure.Jit(t ->
+            new MetaCall(codMeta, metaArgs.appended(t)))));
           return new Jdg.Default(AppTerm.make(acc.wellTyped(), argJdg.wellTyped()), cod);
         }
         case Term otherwise -> throw new ExprTycker.NotPi(otherwise);

--- a/base/src/main/java/org/aya/tyck/ArgsComputer.java
+++ b/base/src/main/java/org/aya/tyck/ArgsComputer.java
@@ -5,12 +5,15 @@ package org.aya.tyck;
 import kala.collection.immutable.ImmutableSeq;
 import kala.collection.mutable.MutableStack;
 import org.aya.generic.term.DTKind;
+import org.aya.generic.term.SortKind;
 import org.aya.syntax.concrete.Expr;
 import org.aya.syntax.core.term.*;
 import org.aya.syntax.core.term.call.ClassCall;
+import org.aya.syntax.core.term.call.MetaCall;
 import org.aya.syntax.core.term.xtt.DimTyTerm;
 import org.aya.syntax.core.term.xtt.EqTerm;
 import org.aya.syntax.ref.LocalVar;
+import org.aya.syntax.ref.MetaVar;
 import org.aya.syntax.telescope.AbstractTele;
 import org.aya.tyck.error.LicitError;
 import org.aya.util.Pair;
@@ -32,7 +35,7 @@ public class ArgsComputer {
   private int paramIx = 0;
   private @Nullable Term firstTy = null;
   private final @NotNull Term @NotNull [] result;
-  private @NotNull Param param;
+  private Param param;
 
   public ArgsComputer(
     @NotNull ExprTycker tycker,
@@ -61,7 +64,10 @@ public class ArgsComputer {
     } else return tycker.mockTerm(param, pos);
   }
 
-  static @NotNull Jdg generateApplication(@NotNull ExprTycker tycker, @NotNull ImmutableSeq<Expr.NamedArg> args, Jdg start) throws ExprTycker.NotPi {
+  static @NotNull Jdg generateApplication(
+    @NotNull ExprTycker tycker,
+    @NotNull ImmutableSeq<Expr.NamedArg> args, Jdg start
+  ) throws ExprTycker.NotPi {
     return args.foldLeftChecked(start, (acc, arg) -> {
       if (arg.name() != null || !arg.explicit()) tycker.fail(new LicitError.BadNamedArg(arg));
       switch (tycker.whnf(acc.type())) {
@@ -72,6 +78,20 @@ public class ArgsComputer {
         case EqTerm eq -> {
           var wellTy = tycker.inherit(arg.arg(), DimTyTerm.INSTANCE).wellTyped();
           return new Jdg.Default(eq.makePApp(acc.wellTyped(), wellTy), eq.appA(wellTy));
+        }
+        case MetaCall(var ref, var metaArgs) -> {
+          var req = ref.req();
+          if (req == MetaVar.Misc.Whatever) req = MetaVar.Misc.IsType;
+          else if (req instanceof MetaVar.OfType(var expected)) {
+            if (!(tycker.whnf(expected) instanceof SortTerm sort)
+              || sort.kind() == SortKind.ISet) {
+              throw new ExprTycker.NotPi(acc.type());
+            }
+          }
+          var argJdg = tycker.synthesize(arg.arg());
+          var codMeta = new MetaVar(ref.name() + "_cod", ref.pos(), ref.ctxSize() + 1, req, false);
+          var cod = new MetaCall(codMeta, metaArgs.appended(argJdg.wellTyped()));
+          return new Jdg.Default(AppTerm.make(acc.wellTyped(), argJdg.wellTyped()), cod);
         }
         case Term otherwise -> throw new ExprTycker.NotPi(otherwise);
       }

--- a/base/src/main/java/org/aya/tyck/StmtTycker.java
+++ b/base/src/main/java/org/aya/tyck/StmtTycker.java
@@ -6,6 +6,7 @@ import kala.collection.immutable.ImmutableSeq;
 import kala.control.Either;
 import kala.control.Option;
 import org.aya.generic.Modifier;
+import org.aya.generic.term.DTKind;
 import org.aya.generic.term.SortKind;
 import org.aya.pretty.doc.Doc;
 import org.aya.primitive.PrimFactory;
@@ -17,7 +18,6 @@ import org.aya.syntax.core.def.*;
 import org.aya.syntax.core.pat.Pat;
 import org.aya.syntax.core.pat.PatToTerm;
 import org.aya.syntax.core.term.*;
-import org.aya.generic.term.DTKind;
 import org.aya.syntax.core.term.call.ClassCall;
 import org.aya.syntax.core.term.call.DataCall;
 import org.aya.syntax.core.term.xtt.DimTyTerm;
@@ -72,12 +72,12 @@ public record StmtTycker(
             var signature = fnRef.signature;
             // In the ordering, we guarantee that expr bodied fn are always checked as a whole
             assert tycker != null;
-            var result = tycker.inherit(expr, tycker.whnf(signature.result().instantiateTeleVar(teleVars.view())))
-              // we still need to bind [result.type()] in case it was a hole
-              .bindTele(teleVars.view());
+            var expectedType = tycker.whnf(signature.result().instantiateTeleVar(teleVars.view()));
+            var result = tycker.inherit(expr, expectedType).wellTyped();
             tycker.solveMetas();
+            var resultTerm = tycker.zonk(result).bindTele(teleVars.view());
             fnRef.signature = fnRef.signature.descent(tycker::zonk);
-            yield factory.apply(Either.left(tycker.zonk(result.wellTyped())));
+            yield factory.apply(Either.left(resultTerm));
           }
           case FnBody.BlockBody(var clauses, var elims, _) -> {
             assert elims != null;

--- a/base/src/main/java/org/aya/tyck/StmtTycker.java
+++ b/base/src/main/java/org/aya/tyck/StmtTycker.java
@@ -72,7 +72,7 @@ public record StmtTycker(
             var signature = fnRef.signature;
             // In the ordering, we guarantee that expr bodied fn are always checked as a whole
             assert tycker != null;
-            var expectedType = tycker.whnf(signature.result().instantiateTeleVar(teleVars.view()));
+            var expectedType = signature.result().instantiateTeleVar(teleVars.view());
             var result = tycker.inherit(expr, expectedType).wellTyped();
             tycker.solveMetas();
             var resultTerm = tycker.zonk(result).bindTele(teleVars.view());

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -38,9 +38,12 @@ import java.util.function.UnaryOperator;
 public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problematic, Stateful {
   public record TyckResult(
     @NotNull ImmutableSeq<Pat.Preclause<Term>> clauses,
-    @NotNull ImmutableSeq<Term.Matching> wellTyped,
     boolean hasLhsError
-  ) { }
+  ) {
+    public @NotNull ImmutableSeq<Term.Matching> wellTyped() {
+      return clauses.flatMap(Pat.Preclause::lift);
+    }
+  }
 
   /**
    * @param paramSubst substitution for parameter, in the same order as parameter.
@@ -116,15 +119,10 @@ public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problemati
     rhsResult = rhsResult.map(preclause -> new Pat.Preclause<>(
       preclause.sourcePos(),
       preclause.pats().map(p -> p.descentTerm(exprTycker::zonk)),
-      preclause.bindCount(),
-      preclause.expr() == null ? null : preclause.expr().descent((_, t) -> exprTycker.zonk(t))
+      preclause.bindCount(), preclause.expr()
     ));
 
-    return new TyckResult(
-      rhsResult,
-      rhsResult.flatMap(Pat.Preclause::lift),
-      lhsError
-    );
+    return new TyckResult(rhsResult, lhsError);
   }
 
   @Override public @NotNull Reporter reporter() { return exprTycker.reporter; }

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -111,7 +111,6 @@ public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problemati
   ) {
     var lhsError = lhsResults.anyMatch(LhsResult::hasError);
     var rhsResult = lhsResults.map(x -> checkRhs(vars, x));
-    exprTycker.solveMetas();
 
     // inline terms in rhsResult
     rhsResult = rhsResult.map(preclause -> new Pat.Preclause<>(
@@ -206,6 +205,8 @@ public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problemati
         result.addLocalLet(teleBinds, exprTycker);
         // now exprTycker has all substitutions that PatternTycker introduced.
         wellBody = exprTycker.inherit(bodyExpr, result.type).wellTyped();
+        exprTycker.solveMetas();
+        wellBody = exprTycker.zonk(wellBody);
 
         // bind all pat bindings
         var patBindTele = Pat.collectVariables(result.clause.pats().view()).component1();

--- a/cli-impl/src/test/resources/negative/GoalAndMeta.txt
+++ b/cli-impl/src/test/resources/negative/GoalAndMeta.txt
@@ -44,29 +44,6 @@ In file $FILE:5:17 ->
   5 │   def test => some 114514
     │                    ╰────╯
 
-Error: Unsolved meta A
-       in `?A`
-       in `114514`
-       in `some 114514`
-
-In file $FILE:5:17 ->
-
-  3 │   open inductive Option (A : Type)
-  4 │     | some A
-  5 │   def test => some 114514
-    │                    ╰────╯
-
-Error: Unsolved meta A
-       in `114514`
-       in `some 114514`
-
-In file $FILE:5:17 ->
-
-  3 │   open inductive Option (A : Type)
-  4 │     | some A
-  5 │   def test => some 114514
-    │                    ╰────╯
-
 Error: Unable to solve the type of this literal:
          114514
        I'm confused about the following candidates:
@@ -83,7 +60,7 @@ Error: Unsolved meta A
        in `Option ?A`
        in `Option ?A`
 
-5 error(s), 0 warning(s).
+3 error(s), 0 warning(s).
 Let's learn from that.
 
 Daylily:
@@ -204,16 +181,6 @@ Error: Unsolved meta _
        in `?_`
        in `[unit]`
 
-In file $FILE:6:4 ->
-
-  4 │   
-  5 │   def good : List Unit => [ ]
-  6 │   def bad => [ unit ]
-    │       ╰─╯
-
-Error: Unsolved meta _
-       in `[unit]`
-
 In file $FILE:6:11 ->
 
   4 │   
@@ -235,7 +202,7 @@ In file $FILE:6:4 ->
 
 Error: Unsolved meta _
 
-4 error(s), 0 warning(s).
+3 error(s), 0 warning(s).
 Let's learn from that.
 
 NonPattern:

--- a/cli-impl/src/test/resources/negative/GoalAndMeta.txt
+++ b/cli-impl/src/test/resources/negative/GoalAndMeta.txt
@@ -35,17 +35,6 @@ In file $FILE:5:17 ->
     │                    ╰────╯
 
 Error: Unsolved meta A
-       in `Option ?A`
-       in `Option ?A`
-
-In file $FILE:5:17 ->
-
-  3 │   open inductive Option (A : Type)
-  4 │     | some A
-  5 │   def test => some 114514
-    │                    ╰────╯
-
-Error: Unsolved meta A
        in `some 114514`
 
 In file $FILE:5:17 ->
@@ -82,6 +71,17 @@ Error: Unable to solve the type of this literal:
          114514
        I'm confused about the following candidates:
          `Nat`, `Nat2`
+
+In file $FILE:5:17 ->
+
+  3 │   open inductive Option (A : Type)
+  4 │     | some A
+  5 │   def test => some 114514
+    │                    ╰────╯
+
+Error: Unsolved meta A
+       in `Option ?A`
+       in `Option ?A`
 
 5 error(s), 0 warning(s).
 Let's learn from that.
@@ -201,15 +201,6 @@ In file $FILE:6:4 ->
     │       ╰─╯
 
 Error: Unsolved meta _
-
-In file $FILE:6:4 ->
-
-  4 │   
-  5 │   def good : List Unit => [ ]
-  6 │   def bad => [ unit ]
-    │       ╰─╯
-
-Error: Unsolved meta _
        in `?_`
        in `[unit]`
 
@@ -234,6 +225,15 @@ Error: Unable to solve the type of this literal:
          [unit]
        I'm confused about the following candidates:
          `List`, `List2`
+
+In file $FILE:6:4 ->
+
+  4 │   
+  5 │   def good : List Unit => [ ]
+  6 │   def bad => [ unit ]
+    │       ╰─╯
+
+Error: Unsolved meta _
 
 4 error(s), 0 warning(s).
 Let's learn from that.

--- a/cli-impl/src/test/resources/success/src/Test.aya
+++ b/cli-impl/src/test/resources/success/src/Test.aya
@@ -51,3 +51,8 @@ module PullRequest1171 {
   def etaTestR (t : Sig Nat ** Nat) : t = (t.1 , t.2) => refl
   def etaTestL (t : Sig Nat ** Nat) : (t.1 , t.2) = t => refl
 }
+
+module Issue1181 {
+  def how {A B : Type} (f : A -> B) (a : A) : B => f a
+  def what {A B : Type} (a : A) (f : A -> B) : B => how (fn g => g a) f
+}

--- a/note/jit.md
+++ b/note/jit.md
@@ -1,0 +1,154 @@
+# Jit Specification
+
+## Jit Unit
+
+Jit Unit 是一个需要被编译的单位, 它包括:
+
+* 任何 Def
+* 顶级模块/子模块
+
+## Class Structure
+
+所有 Jit Unit 都会被编译成一个 class, 它们之间的包含结构通过嵌套类的方式保留.
+
+示例: `class-structure.aya`
+
+```aya
+inductive Nat
+| O
+| S Nat
+
+def inc (n : Nat) : Nat => S n 
+```
+
+编译输出: `class$45structure.java`
+
+```java
+public final class class$45structure {
+  public static final class class$45structure$Nat extends JitData {
+    public static final class class$45structure$Nat$O extends JitCon { /* ... */
+    }
+
+    public static final class class$45structure$Nat$S extends JitCon { /* ... */
+    }
+    // ...
+  }
+
+  public static final class class$45structure$inc extends JitFn { /* ... */
+  }
+}
+```
+
+> 关于编译结果中的类名部分，参见 [Name Mapping](#name-mapping) 章节
+
+任何 Def 对应的类都需要继承对应的基类, 如 `JitData` 对应归纳数据类型, `JitCon` 对应构造子.
+
+## Metadata
+
+除了基类要求实现的各种方法, 这些类还需要带有 `@CompiledAya` 注解, 以提供其他信息, 如:
+
+* 在 aya 中的符号名
+* shape 信息
+* 结合性信息
+* 等等...
+
+## Name Mapping
+
+> ![WARNING]
+> 这个章节仅适用于编译到 Java 源代码的编译器实现.
+
+由于需要在 Java 编译器的限制, 并不是所有合法的 aya 符号名在 Java 中都合法, 因此需要对符号名进行转义.
+
+* 对于是关键字的符号名, 将它转义成 `_$<keyword>`. 如 `new` 转义成 `_$new`
+* 对于符号名中的每个 `$`, 将它转义成 `$$`.
+* 对于符号名中的每个不合法的 Java 符号名 code point, 将它转义成 `$<code point>`. 如 `+` 转义成 `$43`.
+
+## 模式匹配
+
+变量:
+
+* `Term[] result` 保存模式匹配中的 binding 所匹配的 term.
+* `int matchState` 保存模式匹配的状态, 如果这个状态大于 0, 则表示已经匹配到对应的 clause.
+* `boolean subMatchState` 保存局部模式匹配的状态, 有些情况下需要尝试 fastpath, 如 `Pat.ShapedInt` 会先尝试将 term 转换为
+  `IntegerTerm`,
+  如果失败再进行朴素匹配. 此时, 如果 fastpath 匹配成功, 则 `subMathState` 会被设置为 `true`, 从而跳过朴素匹配.
+
+示例: `nat.aya`
+
+```aya
+def plus Nat Nat : Nat
+| 0, b => b
+| S a, b => S (plus a b)
+```
+
+编译输出: `nat.java`
+
+```java
+import java.util.function.Supplier;
+
+// 忽略不重要的内容
+
+public Term invoke(Supplier<Term> onStuck, Term arg0, Term arg1) {
+  Term[] result = new Term[2];    // 所有 clause 中, 最大的 binding 数量是 2
+  int matchState = 0;
+  boolean subMatchState = false;
+
+  // 用于从代码块中跳出
+  do {
+    // 第一个 clause
+    // 匹配: 0
+    subMatchState = false;
+    if (arg0 instanceof IntegerTerm var0) {
+      if (0 = arg0.repr()) {
+        subMatchState = true;
+      }
+    }
+
+    if (!subMatchState) {
+      if (arg0 instanceof ConCallLike var0) {
+        if (Nat$O.INSTANCE.ref() == var0.ref()) {
+          // 匹配: b
+          // b 是一个 Pat.Bind, 匹配总是成功
+          result[0] = var1;
+          matchState = 1;
+        } else {
+          matchState = 0;   // mismatch
+        }
+      } else {
+        matchState = -1;    // stuck
+      }
+    }
+
+    // 第一个 clause 结束, 检查是否匹配成功
+    if (matchState > 0) break;
+
+    // 省略第二个 clause 的匹配
+    // ...
+  } while (false);
+
+  // 根据 matchState 决定行为
+  switch (matchState) {
+    case -1 -> {
+      return onStuck.get();
+    }
+    case 0 -> {
+      // 由于有 pattern classifier, 这个 case 是不可达的.
+      return Panic.unreachable();
+    }
+    case 1 -> {
+      // | 0, b => b
+      return result[0];
+    }
+    case 2 -> {
+      // | S a, b => S (plus a b)
+      return new ConCall(Nat$S.INSTANCE, 0, ImmutableSeq.of(plus.invoke(
+        () -> new FnCall(/* ... */),
+        result[0], result[1]
+      )));
+    }
+    default -> {
+      return Panic.unreachable();
+    }
+  }
+}
+```

